e2e-tests/src/test/button_game/mod.rs

```diff
@@ -5,18 +5,115 @@ use anyhow::Result;
 use assert2::{assert, let_assert};
 use helpers::sign;
 use log::info;
+use sp_core::Pair;
 
 use crate::{
     test::button_game::helpers::{
-        assert_recv, assert_recv_id, refute_recv_id, setup_button_test, wait_for_death,
-        ButtonTestContext,
+        assert_recv, assert_recv_id, mega, refute_recv_id, setup_button_test, setup_dex_test,
+        wait_for_death, ButtonTestContext, DexTestContext,
     },
     Config,
 };
 
 mod contracts;
 mod helpers;
 
+/// Test trading on simple_dex.
+///
+/// The scenario does the following (given 3 tokens A, B, C):
+///
+/// 1. Enables A <-> B, and A -> C swaps.
+/// 2. Adds (A, 2000M), (B, 5000M), (C, 10000M) of liquidity.
+/// 3. Makes a swap A -> B and then B -> A for the amount of B received in the first swap.
+/// 4. Checks that the price after the two swaps is the same as before (with a dust allowance of 1 for rounding).
+/// 5. Checks that it's possible to make an A -> C swap, but impossible to make a C -> A swap.
+pub fn simple_dex(config: &Config) -> Result<()> {
+    let DexTestContext {
+        conn,
+        authority,
+        account,
+        dex,
+        token1,
+        token2,
+        token3,
+        mut events,
+    } = setup_dex_test(config)?;
+    let authority_conn = &sign(&conn, &authority);
+    let account_conn = &sign(&conn, &account);
+    let token1 = token1.as_ref();
+    let token2 = token2.as_ref();
+    let token3 = token3.as_ref();
+    let dex = dex.as_ref();
+
+    dex.add_swap_pair(authority_conn, token1.into(), token2.into())?;
+    assert_recv_id(&mut events, "SwapPairAdded");
+
+    dex.add_swap_pair(authority_conn, token2.into(), token1.into())?;
+    assert_recv_id(&mut events, "SwapPairAdded");
+
+    dex.add_swap_pair(authority_conn, token1.into(), token3.into())?;
+    assert_recv_id(&mut events, "SwapPairAdded");
+
+    token1.mint(authority_conn, &authority.public().into(), mega(3000))?;
+    token2.mint(authority_conn, &authority.public().into(), mega(5000))?;
+    token3.mint(authority_conn, &authority.public().into(), mega(10000))?;
+
+    token1.approve(authority_conn, &dex.into(), mega(3000))?;
+    token2.approve(authority_conn, &dex.into(), mega(5000))?;
+    token3.approve(authority_conn, &dex.into(), mega(10000))?;
+    dex.deposit(
+        authority_conn,
+        &[
+            (token1, mega(3000)),
+            (token2, mega(5000)),
+            (token3, mega(10000)),
+        ],
+    )?;
+
+    let more_than_liquidity = mega(1_000_000);
+    assert!(dex
```

Comment:
would be nice to add custom messages by assertions (eg "should fail due to not enough liquidity")

Reply:
:+1: